src/pool/pool_disjoint.c

```diff
@@ -0,0 +1,1163 @@
+    }
+}
+
+bucket_t *disjoint_pool_find_bucket(disjoint_pool_t *pool, size_t size) {
+    size_t calculated_idx = size_to_idx(pool, size);
+
+#ifndef NDEBUG
+    // debug check
+    bucket_t *bucket = pool->buckets[calculated_idx];
+    assert(bucket->size >= size);
+    (void)bucket;
+
+    if (calculated_idx > 0) {
+        bucket_t *bucket_prev = pool->buckets[calculated_idx - 1];
+        assert(bucket_prev->size < size);
+        (void)bucket_prev;
+    }
+#endif // NDEBUG
+
+    return pool->buckets[calculated_idx];
+}
+
+void bucket_print_stats(bucket_t *bucket, bool *title_printed,
+                        const char *label) {
+    if (bucket->alloc_count) {
+        if (!*title_printed) {
+            LOG_DEBUG("\"%s\" pool memory statistics", label);
+            LOG_DEBUG("%14s %12s %12s %18s %20s %21s", "Bucket Size", "Allocs",
+                      "Frees", "Allocs from Pool", "Peak Slabs in Use",
+                      "Peak Slabs in Pool");
+            *title_printed = true;
+        }
+        LOG_DEBUG("%14zu %12zu %12zu %18zu %20zu %21zu", bucket->size,
+                  bucket->alloc_count, bucket->free_count,
+                  bucket->alloc_pool_count, bucket->max_slabs_in_use,
+                  bucket->max_slabs_in_pool);
+    }
+}
+
+void disjoint_pool_print_stats(disjoint_pool_t *pool, bool *title_printed,
+                               size_t *high_bucket_size,
+                               size_t *high_peak_slabs_in_use,
+                               const char *mt_name) {
+    *high_bucket_size = 0;
+    *high_peak_slabs_in_use = 0;
+    for (size_t i = 0; i < pool->buckets_num; i++) {
+        bucket_t *bucket = pool->buckets[i];
+        bucket_print_stats(bucket, title_printed, mt_name);
+        *high_peak_slabs_in_use =
+            utils_max(bucket->max_slabs_in_use, *high_peak_slabs_in_use);
+        if (bucket->alloc_count) {
+            *high_bucket_size =
+                utils_max(bucket_slab_alloc_size(bucket), *high_bucket_size);
+        }
+    }
+}
+
+void *disjoint_pool_allocate(disjoint_pool_t *pool, size_t size,
+                             bool *from_pool) {
+    if (size == 0) {
+        return NULL;
+    }
+
+    void *ptr = NULL;
+
+    if (size > pool->params.max_poolable_size) {
+        umf_result_t ret =
+            umfMemoryProviderAlloc(pool->provider, size, 0, &ptr);
+        if (ret != UMF_RESULT_SUCCESS) {
+            TLS_last_allocation_error = ret;
+            return NULL;
+        }
+
+        annotate_memory_undefined(ptr, size);
+        return ptr;
+    }
+
+    *from_pool = false;
+    bucket_t *bucket = disjoint_pool_find_bucket(pool, size);
+
+    utils_mutex_lock(&bucket->bucket_lock);
+
+    if (size > bucket_chunk_cut_off(bucket)) {
+        ptr = bucket_get_slab(bucket, from_pool);
+    } else {
+        ptr = bucket_get_chunk(bucket, from_pool);
+    }
+
+    if (pool->params.pool_trace > 1) {
+        bucket_count_alloc(bucket, from_pool);
```

Comment:
This should probably only be done if(ptr)

Reply:
done